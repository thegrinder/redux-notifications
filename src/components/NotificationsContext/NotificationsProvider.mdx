---
name: NotificationsProvider
---

import { Props } from 'docz';
import { NotificationsProvider } from '.';

# NotificationsProvider

NotificationsProvider component is just a context provider that takes a few configuration props.
The most important is renderNotification that allows for maximum flexibility when creating your
own custom notification component.

<Props of={NotificationsProvider} />
