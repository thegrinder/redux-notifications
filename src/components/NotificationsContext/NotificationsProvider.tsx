import React, {
  useCallback,
  useState,
  useRef,
  useEffect,
  useMemo,
  ReactNode,
} from 'react';
import Container from '../Container/Container';
import { Notification } from '../Notification/Notification';
import { Position } from '../../helpers/utils';
import { NotificationsContext } from './NotificationContext';

export type RemoveNotification = ({
  removeNotification,
  payload,
}: {
  removeNotification: () => void;
  payload: any;
}) => ReactNode;

type NotificationsProviderProps = {
  /** Children */
  children: ReactNode;
  /** Render prop which passes down removeNotification function and notification payload */
  renderNotification: RemoveNotification;
  /** Fixed position where all notifications are displayed */
  position?: Position;
  /** Duration of the show and hide animations in miliseconds */
  animationDuration?: number;
  /** Animation timing function / cubic-bezier */
  animationEasing?: string;
  /** Time in miliseconds after which the notification is automatically dismissed */
  dismissAfter?: number;
};

export const NotificationsProvider = ({
  children,
  renderNotification,
  position = ['40px', '40px', 'auto', 'auto'],
  animationDuration = 400,
  animationEasing = 'ease',
  dismissAfter = 6000,
}: NotificationsProviderProps) => {
  const automaticDismissalTimers = useRef<Record<string, NodeJS.Timeout>>({});
  const manualDismissalTimers = useRef<Record<string, NodeJS.Timeout>>({});
  const appearanceDelayTimers = useRef<Record<string, NodeJS.Timeout>>({});

  const [notifications, updateNotifications] = useState<
    Record<
      string,
      {
        id: string;
        isVisible: boolean;
        payload: any;
      }
    >
  >({});

  const hideNotification = (id: string) =>
    updateNotifications((state) => ({
      ...state,
      [id]: {
        ...state[id],
        isVisible: false,
      },
    }));

  const unmountNotification = (id: string) =>
    updateNotifications((state) =>
      Object.keys(state)
        .filter((notificationId) => notificationId !== id)
        .reduce(
          (acc, notificationId) => ({
            ...acc,
            [notificationId]: state[notificationId],
          }),
          {}
        )
    );

  const removeNotification = useCallback(
    (id) => () => {
      clearTimeout(automaticDismissalTimers.current[id]);
      clearTimeout(appearanceDelayTimers.current[id]);
      hideNotification(id);
      manualDismissalTimers.current[id] = setTimeout(() => {
        unmountNotification(id);
      }, animationDuration);
    },
    [animationDuration]
  );

  const mountNotification = useCallback(
    (id, payload) => {
      updateNotifications((state) => {
        automaticDismissalTimers.current[id] = setTimeout(() => {
          removeNotification(id)();
        }, dismissAfter);
        return {
          ...state,
          [id]: {
            id,
            isVisible: false,
            payload,
          },
        };
      });
    },
    [dismissAfter, removeNotification]
  );

  const showNotification = (id: string) =>
    updateNotifications((state) => ({
      ...state,
      [id]: {
        ...state[id],
        isVisible: true,
      },
    }));

  const addNotification = useCallback(
    (payload = {}) => {
      const id = Date.now().toString();
      mountNotification(id, payload);
      appearanceDelayTimers.current[id] = setTimeout(() => {
        showNotification(id);
      }, 100);
    },
    [mountNotification]
  );

  const clearAllTimeouts = () => {
    [
      ...Object.values(manualDismissalTimers.current),
      ...Object.values(automaticDismissalTimers.current),
      ...Object.values(appearanceDelayTimers.current),
    ].forEach((timeout) => clearTimeout(timeout));
  };

  useEffect(() => clearAllTimeouts, []);

  const value = useMemo(
    () => ({
      addNotification,
      removeNotification,
    }),
    [removeNotification, addNotification]
  );

  return (
    <NotificationsContext.Provider value={value}>
      <Container position={position}>
        {Object.values(notifications).map(({ id, payload, isVisible }) => (
          <Notification
            key={id}
            position={position}
            animationDuration={animationDuration}
            animationEasing={animationEasing}
            isVisible={isVisible}
          >
            {renderNotification({
              removeNotification: removeNotification(id),
              payload,
            })}
          </Notification>
        ))}
      </Container>
      {children}
    </NotificationsContext.Provider>
  );
};
