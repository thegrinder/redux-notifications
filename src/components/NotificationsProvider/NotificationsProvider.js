import React, {
  createContext,
  useCallback,
  useState,
  useRef,
  useEffect,
  useMemo,
} from 'react';
import PropTypes from 'prop-types';
import Container from '../Container/Container';
import Notification from '../Notification/Notification';

export const NotificationsContext = createContext({});

const propTypes = {
  /** Children */
  children: PropTypes.node.isRequired,
  /** Render prop which passes down removeNotification function and notification payload */
  renderNotification: PropTypes.func.isRequired,
  /** Fixed position where all notifications are displayed */
  position: PropTypes.array,
  /** Duration of the show and hide animations in miliseconds */
  animationDuration: PropTypes.number,
  /** Animation timing function / cubic-bezier */
  animationEasing: PropTypes.string,
  /** Time in miliseconds after which the notification is automatically dismissed */
  dismissAfter: PropTypes.number,
};

const defaultProps = {
  position: ['40px', '40px', 'auto', 'auto'],
  animationDuration: 400,
  animationEasing: 'ease',
  dismissAfter: 6000,
};

const NotificationsProvider = ({
  children,
  renderNotification,
  position,
  animationDuration,
  animationEasing,
  dismissAfter,
}) => {
  const automaticDismissalTimers = useRef({});
  const manualDismissalTimers = useRef({});
  const appearanceDelayTimers = useRef({});

  const [notifications, updateNotifications] = useState({});

  const hideNotification = (id) =>
    updateNotifications((state) => ({
      ...state,
      [id]: {
        ...state[id],
        isVisible: false,
      },
    }));

  const unmountNotification = (id) =>
    updateNotifications((state) =>
      Object.keys(state)
        .filter((notificationId) => notificationId !== id)
        .reduce(
          (acc, notificationId) => ({
            ...acc,
            [notificationId]: state[notificationId],
          }),
          {}
        )
    );

  const removeNotification = useCallback(
    (id) => () => {
      clearTimeout(automaticDismissalTimers.current[id]);
      clearTimeout(appearanceDelayTimers.current[id]);
      hideNotification(id);
      manualDismissalTimers.current[id] = setTimeout(() => {
        unmountNotification(id);
      }, animationDuration);
    },
    [animationDuration]
  );

  const mountNotification = useCallback(
    (id, payload) => {
      updateNotifications((state) => {
        automaticDismissalTimers.current[id] = setTimeout(() => {
          removeNotification(id)();
        }, dismissAfter);
        return {
          ...state,
          [id]: {
            id,
            isVisible: false,
            payload,
          },
        };
      });
    },
    [dismissAfter, removeNotification]
  );

  const showNotification = (id) =>
    updateNotifications((state) => ({
      ...state,
      [id]: {
        ...state[id],
        isVisible: true,
      },
    }));

  const addNotification = useCallback(
    (payload = {}) => {
      const id = Date.now().toString();
      mountNotification(id, payload);
      appearanceDelayTimers.current[id] = setTimeout(() => {
        showNotification(id);
      }, 100);
    },
    [mountNotification]
  );

  const clearAllTimeouts = () => {
    [
      ...Object.values(manualDismissalTimers.current),
      ...Object.values(automaticDismissalTimers.current),
      ...Object.values(appearanceDelayTimers.current),
    ].forEach((timeout) => clearTimeout(timeout));
  };

  useEffect(() => clearAllTimeouts, []);

  const value = useMemo(
    () => ({
      addNotification,
      removeNotification,
    }),
    [removeNotification, addNotification]
  );

  return (
    <NotificationsContext.Provider value={value}>
      <Container position={position}>
        {Object.values(notifications).map(({ id, payload, isVisible }) => (
          <Notification
            key={id}
            position={position}
            animationDuration={animationDuration}
            animationEasing={animationEasing}
            dismissAfter={dismissAfter}
            isVisible={isVisible}
          >
            {renderNotification({
              removeNotification: removeNotification(id),
              payload,
            })}
          </Notification>
        ))}
      </Container>
      {children}
    </NotificationsContext.Provider>
  );
};

NotificationsProvider.propTypes = propTypes;
NotificationsProvider.defaultProps = defaultProps;

export default NotificationsProvider;
