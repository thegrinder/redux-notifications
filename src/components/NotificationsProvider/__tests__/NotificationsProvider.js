import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { render, cleanup, fireEvent, act } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import NotificationsProvider, {
  NotificationsContext,
} from '../NotificationsProvider';

const TestChild = () => {
  const { showNotification } = useContext(NotificationsContext);
  return (
    <button type="button" onClick={() => showNotification()}>
      show notification
    </button>
  );
};

const children = <TestChild />;

const Notification = ({ removeNotification }) => (
  <div>
    <button type="button" onClick={() => removeNotification()}>
      notification
    </button>
  </div>
);

Notification.propTypes = {
  removeNotification: PropTypes.func.isRequired,
};

const requiredProps = {
  renderNotification: Notification,
};

const animationDuration = 400;

const optionalProps = {
  animationDuration,
  position: ['40px', '40px', 'auto', 'auto'],
  animationEasing: 'ease',
};

const renderComponent = (props = {}) =>
  render(
    <NotificationsProvider {...requiredProps} {...props}>
      {children}
    </NotificationsProvider>
  );

describe('<NotificationsProvider />', () => {
  afterEach(cleanup);

  it('should render correctly', () => {
    const {
      container: { firstChild },
    } = renderComponent();
    expect(firstChild).toBeDefined();
    expect(firstChild).toMatchSnapshot();
  });

  it('should have children', () => {
    const { container, getByText } = renderComponent();
    const childrenElement = getByText('show notification');
    expect(container).toContainElement(childrenElement);
  });

  it('should render correctly with custom props', () => {
    const {
      container: { firstChild },
    } = renderComponent(optionalProps);
    expect(firstChild).toMatchSnapshot();
  });

  it('should properly handle opening and manually closing notifications', () => {
    jest.useFakeTimers();

    const { container, getAllByText, getByText } = renderComponent();
    const button = getByText('show notification');

    act(() => {
      fireEvent.click(button);
    });

    act(() => {
      fireEvent.click(button);
    });

    expect(getAllByText('notification').length).toEqual(2);
    expect(container).toMatchSnapshot();
    act(() => {
      fireEvent.click(getAllByText('notification')[0]);
    });
    act(() => {
      jest.advanceTimersByTime(animationDuration);
    });
    expect(getAllByText('notification').length).toEqual(1);
  });

  it('should automatically dismiss notifications', () => {
    jest.useFakeTimers();

    const { queryAllByText, getByText } = renderComponent();
    const button = getByText('show notification');

    act(() => {
      fireEvent.click(button);
    });

    act(() => {
      fireEvent.click(button);
    });

    expect(queryAllByText('notification').length).toEqual(2);
    act(() => {
      jest.runAllTimers();
    });
    expect(queryAllByText('notification').length).toEqual(0);
  });
});
