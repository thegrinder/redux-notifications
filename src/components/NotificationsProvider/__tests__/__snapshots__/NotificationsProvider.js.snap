// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotificationsProvider /> should properly handle opening and manually closing notifications 1`] = `
<div>
  <div
    style="position: fixed; display: flex; z-index: 999999; height: 0px; flex-direction: column; top: 40px; right: 40px;"
  >
    <div
      style="position: relative; z-index: -1; transition-property: margin-top, opacity; transition-duration: 0.4s; transition-timing-function: ease; margin-top: 0px; opacity: 0;"
    >
      <div>
        <button
          type="button"
        >
          notification
        </button>
      </div>
    </div>
    <div
      style="position: relative; z-index: -1; transition-property: margin-top, opacity; transition-duration: 0.4s; transition-timing-function: ease; margin-top: 0px; opacity: 0;"
    >
      <div>
        <button
          type="button"
        >
          notification
        </button>
      </div>
    </div>
  </div>
  <button
    type="button"
  >
    show notification
  </button>
</div>
`;

exports[`<NotificationsProvider /> should render correctly 1`] = `
<div
  style="position: fixed; display: flex; z-index: 999999; height: 0px; flex-direction: column; top: 40px; right: 40px;"
/>
`;

exports[`<NotificationsProvider /> should render correctly with custom props 1`] = `
<div
  style="position: fixed; display: flex; z-index: 999999; height: 0px; flex-direction: column; top: 40px; right: 40px;"
/>
`;
